### ISSUE TITLE
dnsmasq-full dnssec does not get a sighup after time is slewed

### ISSUE THREAD

**Issue by dtaht** [dtaht/dnsmasq-lede#5]
I need to see what other people's scripts look like but, on boot, I have this happen.

Tue Jan 24 13:55:22 2017 daemon.info dnsmasq[2902]: DNSSEC validation enabled
Tue Jan 24 13:55:22 2017 daemon.info dnsmasq[2902]: DNSSEC signature timestamps not checked until first cache reload
Tue Jan 24 13:55:22 2017 daemon.info dnsmasq-dhcp[2902]: DHCP, IP range 172.26.64.20 -- 172.26.64.254, lease time 12h
Tue Jan 24 13:55:22 2017 daemon.warn dnsmasq[2902]: ignoring nameserver 172.26.64.1 - local interface
Tue Jan 24 13:55:22 2017 daemon.info dnsmasq[2902]: using nameserver 172.26.20.1#53 for domain lorna-side.taht.net (no DNSSEC)
Tue Jan 24 13:55:22 2017 daemon.info dnsmasq[2902]: using nameserver 172.26.16.1#53 for domain gw.taht.net (no DNSSEC)
Tue Jan 24 13:55:22 2017 daemon.info dnsmasq[2902]: using local addresses only for domain lan
Tue Jan 24 13:55:22 2017 daemon.info dnsmasq[2902]: reading /tmp/resolv.conf.auto
Tue Jan 24 13:55:22 2017 daemon.info dnsmasq[2902]: using nameserver 172.26.20.1#53 for domain lorna-side.taht.net (no DNSSEC)
Tue Jan 24 13:55:22 2017 daemon.info dnsmasq[2902]: using nameserver 172.26.16.1#53 for domain gw.taht.net (no DNSSEC)
Tue Jan 24 13:55:22 2017 daemon.info dnsmasq[2902]: using local addresses only for domain lan
Tue Jan 24 13:55:22 2017 daemon.info dnsmasq[2902]: using nameserver 208.201.224.11#53
Tue Jan 24 13:55:22 2017 daemon.info dnsmasq[2902]: using nameserver 208.201.224.33#53
Tue Jan 24 13:55:22 2017 daemon.info dnsmasq[2902]: read /etc/hosts - 5 addresses
Tue Jan 24 13:55:22 2017 daemon.info dnsmasq[2902]: read /tmp/hosts/odhcpd - 0 addresses
Tue Jan 24 13:55:22 2017 daemon.info dnsmasq[2902]: read /tmp/hosts/dhcp.cfg02411c - 2 addresses
Tue Jan 24 13:55:22 2017 daemon.info dnsmasq-dhcp[2902]: read /etc/ethers - 0 addresses
Tue Jan 24 13:55:30 2017 daemon.warn odhcpd[1291]: A default route is present but there is no public prefix on br-lan thus we don't announce a default route!
Tue Jan 24 13:55:38 2017 cron.err crond[1319]: time disparity of -11032040 minutes detected

 but we slewed the time! And I had to manually killall -1 dnsmasq, which then reloaded.

Tue Jan 24 13:55:22 2017 daemon.info dnsmasq-dhcp[2902]: read /etc/ethers - 0 addresses
Tue Jan 24 13:55:30 2017 daemon.warn odhcpd[1291]: A default route is present but there is no public prefix on br-lan thus we don't announce a default route!
Tue Jan 24 13:55:38 2017 cron.err crond[1319]: time disparity of -11032040 minutes detected
Tue Jan 24 14:00:00 2017 cron.info crond[1319]: USER root pid 2914 cmd /sbin/fan_ctrl.sh
Tue Jan 24 14:00:17 2017 daemon.notice netifd: wan (1587): udhcpc: sending renew
Tue Jan 24 14:00:17 2017 daemon.notice netifd: wan (1587): udhcpc: lease of 198.27.248.35 obtained, lease time 600
Tue Jan 24 14:00:20 2017 daemon.info dnsmasq[2902]: now checking DNSSEC signature timestamps
Tue Jan 24 14:00:20 2017 daemon.info dnsmasq[2902]: read /etc/hosts - 5 addresses
Tue Jan 24 14:00:20 2017 daemon.info dnsmasq[2902]: read /tmp/hosts/odhcpd - 0 addresses
Tue Jan 24 14:00:20 2017 daemon.info dnsmasq[2902]: read /tmp/hosts/dhcp.cfg02411c - 2 addresses
Tue Jan 24 14:00:20 2017 daemon.info dnsmasq-dhcp[2902]: read /etc/ethers - 0 addresses
root@linksys-1200ac:~#

**Comment by EricLuehrsen** [comment]
LEDE changed the init / hotplug scripts to use procd triggers to issue a reload. The answer might be there. For packages/net/unbound, I configured hotplug/ntp to just smash it with a whole restart. Its once a boot and before the cache really fills. So reliability wins over fancy.

**Comment by ldir-EDB0** [comment]
I daresay it's too late to find out this info but does/did /var/state/dnsmasqsec exist?

How the mechanism is supposed to work (and works on my system) is:  (busybox) ntpd notices a stratum change and calls /etc/hotplug.d/25-dnsmasqsec.  25-dnsmasqsec checks to see that it's being informed of a stratum change and then looks for the '/var/state/dnsmasqsec' 'time valid' flag file.  If the file does not exist then this is the first time slew to correct time - at which point we create the flag file & tell procd to SIGHUP all dnsmasq instances - SIGHUP tells dnsmasq to enable timestamp checking and flush its DNS cache.

If dnsmasq is restarted completely then /var/state/dnsmasqsec is used as a flag to say whether it should start with dnssec timechecking disabled (until SIGHUP - assumed to be from the above ntpd process) or if it should check timestamps from the get go.

**Comment by dtaht** [comment]
Song of the day. This is so more byzantine than I wanted.

https://www.youtube.com/watch?v=tkplPbd2f60

**Comment by EricLuehrsen** [comment]
Yeah. [repeat unbound above]. But its early in boot and flushes cache anyway, so just rewrite the conf file and hard restart. Guaranteed results.
